lib/6to5/traverse/index.js

```diff
@@ -50,104 +57,139 @@ TraversalContext.prototype.maybeReplace = function (result, obj, key, node) {
+  return this.didStop;
+};
+
+TraversalContext.prototype.visit = function (node, key, opts, scope, state) {
+  var nodes = node[key];
+  if (!nodes) return;
+
+  if (!Array.isArray(nodes)) {
+    return this.visitNode(node, key, opts, scope, node, state);
   }
 
-  // traverse node
-  traverse(node, opts, ourScope);
+  if (nodes.length === 0) {
+    return;
+  }
 
-  // exit
-  if (opts.exit) {
-    result = opts.exit.call(this, node, parent, ourScope);
-    node = this.maybeReplace(result, obj, key, node);
+  for (var k = 0; k < nodes.length; k++) {
+    if (nodes[k] && this.visitNode(nodes, k, opts, scope, node, state))
+      return true;
   }
-};
 
-function traverse(parent, opts, scope) {
-  // falsy node
-  if (!parent) return;
+  if (this.didFlatten) {
+    node[key] = _.flatten(node[key]);
 
-  // array of nodes
-  if (Array.isArray(parent)) {
-    for (var i = 0; i < parent.length; i++)
-      traverse(parent[i], opts, scope);
-    return;
+    if (key === "body") {
+      // we can safely compact this
+      node[key] = _.compact(node[key]);
+    }
   }
+};
 
-  // unknown node type to traverse
-  var keys = t.VISITOR_KEYS[parent.type];
+function traverseNode(node, opts, scope, state) {
+  var keys = t.VISITOR_KEYS[node.type];
   if (!keys) return;
 
-  opts = opts || {};
-  var context = null;
-
+  var context = new TraversalContext();
   for (var j = 0; j < keys.length; j++) {
-    var key = keys[j];
-    var nodes = parent[key];
-    if (!nodes) continue;
-
-    if (Array.isArray(nodes)) {
-      for (var k = 0; k < nodes.length; k++) {
-        context = new TraversalContext(context);
-        context.visit(nodes, k, opts, scope, parent);
-        if (context.didStop) return;
-      }
+    if (context.visit(node, keys[j], opts, scope, state))
+      return;
+  }
+}
 
-      if (context && context.didFlatten) {
-        parent[key] = _.flatten(parent[key]);
+function traverse(parent, opts, scope, state) {
+  // falsy node
+  if (!parent) return;
 
-        if (key === "body") {
-          // we can safely compact this
-          parent[key] = _.compact(parent[key]);
-        }
-      }
-    } else {
-      context = new TraversalContext(context);
-      context.visit(parent, key, opts, scope, parent);
-      if (context.didStop) return;
-    }
+  if (!opts) opts = {};
+  if (!opts.enter) opts.enter = noop;
+  if (!opts.exit) opts.exit = noop;
+
+  // array of nodes
+  if (!Array.isArray(parent)) {
+    traverseNode(parent, opts, scope, state);
+    return;
+  }
+
+  for (var i = 0; i < parent.length; i++) {
+    traverseNode(parent[i], opts, scope, state);
   }
 }
 
 traverse.removeProperties = function (tree) {
   var clear = function (node) {
-    delete node._declarations;
-    delete node.extendedRange;
-    delete node._scopeInfo;
-    delete node.tokens;
-    delete node.range;
-    delete node.start;
-    delete node.end;
-    delete node.loc;
-    delete node.raw;
+    node._declarations = null;
```

Comment:
[`delete` causes switch to slow lookup mode](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#521-objects-that-are-in-hash-table-mode-aka-normalized-objects-dictionary-mode---objects-who-have-a-hash-table-as-a-backing-data-structure-are-not-simple-enumerables)

But I'm afraid `node`s are quite possibly of different shapes. But that's whole another issue because, as I understand, they're not ours (we get them from acorn?)